static/app/utils/profiling/flamegraph/selectNearestFrame.ts

```diff
@@ -0,0 +1,147 @@
+import {FlamegraphFrame} from '../flamegraphFrame';
+
+export type DirectionX = 'left' | 'right';
+export type DirectionY = 'up' | 'down';
+export type Direction = DirectionY | DirectionX;
+/**
+ * selectNearestFrame walks a FlamegraphFrame tree the direction specified and returns the nearest
+ * FlamegraphFrame relative to the starting node.
+ *
+ * below is a gist of the walking algorithm:
+ *
+ *  up    =>
+ *          |> if parent is present, goto parent
+ *          |> if no parent & previousSibling is present, goto maxDepth of previousSibling
+ *          |> if no parent & no previousSibling, no-op
+ *  down  =>
+ *          |> if immediate child is present, goto first child
+ *          |> if no child & walk up to nearest nextSibling
+ *          |> if at max depth & no nextSibling, no-op
+ *  left  =>
+ *          |> walk up to find a node that has a previousSibling
+ *          |> walk down the sibling tree following right branches
+ *          |> return node matching target depth, if max depth found is less than target return
+ *  right => same as left, but we find the right sibling, and follow left branches down until target/max depth is reached
+ */
+export function selectNearestFrame(frame: FlamegraphFrame, direction: Direction) {
+  const targetDepth = frame.depth;
+  const parent = frame.parent;
+  const node = frame;
+
+  if (direction === 'up') {
+    if (parent) {
+      return parent;
+    }
+
+    const leftSibling = getSibling(node, 'left');
+    if (leftSibling) {
+      return leftSibling;
+    }
+
+    return frame;
+  }
+
+  const child = frame.children?.[0];
+  if (direction === 'down') {
+    if (child) {
+      return child;
+    }
+    const sibling = scanForNearestSibling(node, 'right');
+    if (sibling) {
+      return sibling;
+    }
+
+    return frame;
+  }
+
+  const sibling = scanForNearestSibling(node, direction);
+
+  if (sibling) {
+    const nodeWithDepth = scanForNearestFrameWithDepth(sibling, targetDepth, direction);
+    if (nodeWithDepth) {
+      return nodeWithDepth;
+    }
+  }
+
+  return frame;
+}
+
+function scanForNearestSibling(frame: FlamegraphFrame, directionX: DirectionX) {
+  let node = frame;
+  while (node) {
+    let parent = node.parent;
+    const sibling = getSibling(node, directionX);
+
+    if (sibling) {
+      return sibling;
+    }
+
+    if (!parent) {
+      break;
+    }
+
+    node = parent;
+    parent = parent.parent;
+  }
+  return frame;
```

Comment:
This is just returning the fn inpu, shouldn't it return `node` instead?

Reply:
we could... i designed it so that it would treat it as a hard boundary. if we were to return node, we'd jump stacks to the target depth.

moving `right` here would bounce me to a frame in the first stack 
![image](https://user-images.githubusercontent.com/7349258/196805268-5a7fc0dc-c1a4-44f2-ae3a-f8c7785ceb6f.png)